### PR TITLE
NIAD-1961: Encounter Mapper

### PR DESCRIPTION
**Pull request by rorysproule1**
[NIAD-1961](https://gpitbjss.atlassian.net/browse/NIAD-1961)

This ticket brings in functionality for both mapping the ehrComposition -> Encounter ... and then creating the appropriate Lists relevant to this Encounter. The flow for this can be followed in both the Ecnounter and Consultation Structure tabs in the spreadsheet.

NOTE: There will be a future ticket to implement the referencing of related resources as entries in the List resources

### CODE REVIEW

File: gp2gp-translator/src/integrationTest/java/uk/nhs/adaptors/pss/translator/EhrExtractHandlingIT.java [hunk 1]
```diff
@@ -43,8 +43,11 @@
     private static final String NHS_NUMBER_PLACEHOLDER = "{{nhsNumber}}";
     private static final List<String> IGNORED_JSON_PATHS = List.of(
         "id",
-        "entry[0].resource.id",
```
Comment (dawidskainos) [inline gp2gp-translator/src/integrationTest/java/uk/nhs/adaptors/pss/translator/EhrExtractHandlingIT.java:46]: Do you think we could list all the specific entries to ignore rather than ignore [*]? Most of IDs use the value from HL7 XML.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+    private static final String LOCATION_REFERENCE_SUFFIX = "-LOC";
+    private static final String PERFORMER_SYSTEM = "http://hl7.org/fhir/v3/ParticipationType";
+    private static final String PERFORMER_CODE = "PPRF";
+    private static final String PERFORMER_DISPLAY = "primary performer";
+    private static final String RECORDER_SYSTEM = "https://fhir.nhs.uk/STU3/CodeSystem/GPConnect-ParticipantType-1";
+    private static final String RECORDER_CODE = "REC";
+    private static final String RECORDER_DISPLAY = "recorder";
+    private static final String TOPIC_CLASS_CODE = "TOPIC";
+    private static final String CATEGORY_CLASS_CODE = "CATEGORY";
+    private static final String ENCOUNTER_KEY = "encounters";
```
Comment (dawidskainos) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:64]: Perhaps we could use ResourceType enum as key?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    private static final String CONSULTATION_KEY = "consultations";
+    private static final String TOPIC_KEY = "topics";
+    private static final String CATEGORY_KEY = "categories";
+    private static final String IDENTIFIER_SYSTEM = "https://PSSAdaptor/";
+
+    private final CodeableConceptMapper codeableConceptMapper;
+    private final ListMapper listMapper;
+
+    public Map<String, List<? extends DomainResource>> mapEncounters(RCMRMT030101UK04EhrExtract ehrExtract, Patient patient) {
```
Comment (dawidskainos) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:73]: Have you considered EnumMap? ResourceType enum could be the key.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+        List<RCMRMT030101UK04EhrComposition> ehrCompositionList = getEncounterEhrCompositions(ehrExtract);
+
+        ehrCompositionList.forEach(ehrComposition -> {
+            var encounter = mapToEncounter(ehrComposition, patient);
+            var consultation = listMapper.mapToConsultation(ehrExtract, encounter);
+
+            var topicCompoundStatementList = getTopicCompoundStatements(ehrComposition);
+            if (CollectionUtils.isEmpty(topicCompoundStatementList)) {
+                // generate a 'flat' consultation
```
Comment (bbell-kainos) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:89]: I mentioned in another PR if there is a need for comments? I guess its up for debate, although my view for this project in particular would be that they add noise, since they are only ever really for TODOs. Perhaps extract common logic that is commented into a separate private method- eg. generateFlatConsultation?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+    private void setEncounterLocation(Encounter encounter, RCMRMT030101UK04EhrComposition ehrComposition) {
+        if (ehrComposition.getLocation() != null) {
+            var location = new EncounterLocationComponent();
+            location.setLocation(new Reference(LOCATION_REFERENCE_PREFIX
+                + ehrComposition.getId().getRoot() + LOCATION_REFERENCE_SUFFIX));
+
+            encounter.setLocation(List.of(location));
+        }
+    }
+}
```
Comment (bbell-kainos) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:327]: Correct me if I'm wrong but I think the standard is that all Java classes in this project should end with an empty space?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+    private Meta getListMeta() {
+        return new Meta().addProfile(LIST_META_PROFILE);
+    }
+
+    private CodeableConcept getListCoding(String system, String code, String display) {
+        Coding coding = new Coding(system, code, display);
+        return new CodeableConcept(coding);
+    }
+}
```
Comment (bbell-kainos) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java:155]: Extra line needed at end of class?

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/EncounteMapperTest.java [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+
+import lombok.SneakyThrows;
+import uk.nhs.adaptors.pss.translator.service.IdGeneratorService;
+
+@SpringBootTest
+@ExtendWith(MockitoExtension.class)
+public class EncounteMapperTest {
```
Comment (bbell-kainos) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/EncounteMapperTest.java:42]: Spelling mistake on test class name

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/EncounteMapperTest.java [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+        lenient().when(codeableConceptMapper.mapToCodeableConcept(any())).thenReturn(codeableConcept);
+    }
+
+    @SneakyThrows
+    private RCMRMT030101UK04EhrExtract unmarshallEhrExtractElement(String fileName) {
+        return unmarshallFile(getFile("classpath:" + XML_RESOURCES_BASE + fileName), RCMRMT030101UK04EhrExtract.class);
+    }
+}
+
+
```
Comment (bbell-kainos) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/EncounteMapperTest.java:334]: Remove this extra line please

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+            participantList.add(getPerformer(participant2.get()));
+        }
+
+        return participantList;
+    }
+
+    private void setEncounterLocation(Encounter encounter, RCMRMT030101UK04EhrComposition ehrComposition) {
+        if (ehrComposition.getLocation() != null) {
+            var location = new EncounterLocationComponent();
+            location.setLocation(new Reference(LOCATION_REFERENCE_PREFIX
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:321]: Instead of having prefix and suffix as a separate constants and joining them using `+`, you could have one constant `LOCATION_REFERENCE =  "Location/%s-LOC"`, and use `.formatted(ehrComposition.getId().getRoot())` method on it.

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ListMapperTest.java [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+        var topic = setUpTopic();
+        var compoundStatement = setUpCompoundStatement("test-text", "test-display", "20150213152000");
+
+        var category = listMapper.mapToCategory(topic, compoundStatement);
+
+        assertCategory(category, "20150213152000", "test-text");
+    }
+
+    @Test
+    public void testValidFallbackDateCategoryList() {
```
Comment (bbell-kainos) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ListMapperTest.java:143]: Should this be ...FallbackData... ?

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ProcedureRequestMapperTest.java [hunk 11]
```diff
@@ -177,4 +177,3 @@ private RCMRMT030101UK04EhrExtract unmarshallCodeElement(String fileName) {
         return unmarshallFile(getFile("classpath:" + XML_RESOURCES_BASE + fileName), RCMRMT030101UK04EhrExtract.class);
     }
 }
```
Comment (bbell-kainos) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ProcedureRequestMapperTest.java:179]: Extra line needed at end of class?

**Review by bbell-kainos — CHANGES_REQUESTED** [review]

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+
+            var topicCompoundStatementList = getTopicCompoundStatements(ehrComposition);
+            if (CollectionUtils.isEmpty(topicCompoundStatementList)) {
+                // generate a 'flat' consultation
+                var topic = listMapper.mapToTopic(consultation, null);
+                consultation.addEntry(new ListEntryComponent(new Reference(topic)));
+
+                topicList.add(topic);
+            } else {
+                // generate a 'structured' topic list for each child TOPIC CompoundStatement
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:95]: I'd consider moving the whole logic from `else` to a separate method. Then you could name it accordingly and the comment wouldn't be needed anymore. Maybe it could also make sense to create additional methods for `generate a category list for each CATEGORY CompoundStatement` and `generate a topic list if any top level LinkSets exist, for conditions to be referenced from` sections.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+
+import lombok.RequiredArgsConstructor;
+import uk.nhs.adaptors.pss.translator.service.IdGeneratorService;
+import uk.nhs.adaptors.pss.translator.util.DateFormatUtil;
+
+@Service
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class ListMapper {
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java:27]: `ListMapper` name seems generic (as it doesn't say what kind of list it can be used for). If it's only for one kind of ListResource, then maybe `ConsultationListMapper` would be a better name? But if you think we'll later add methods to map other lists, then it's fine :)

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+            .setDateElement(getDate(compoundStatement, consultation))
+            .setOrderedBy(getListCoding(LIST_ORDERED_BY_SYSTEM, LIST_ORDERED_BY_CODE, LIST_ORDERED_BY_DISPLAY))
+            .setMeta(getListMeta())
+            .setId(getTopicId(compoundStatement));
+
+        return topic;
+    }
+
+    private String getTopicId(RCMRMT030101UK04CompoundStatement compoundStatement) {
+        return compoundStatement != null ? compoundStatement.getId().get(0).getRoot() : idGenerator.generateUuid(); // TODO: double check
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java:105]: TODO left here :D

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+            .setDateElement(getDate(compoundStatement, topic))
+            .setOrderedBy(getListCoding(LIST_ORDERED_BY_SYSTEM, LIST_ORDERED_BY_CODE, LIST_ORDERED_BY_DISPLAY))
+            .setMeta(getListMeta())
+            .setId(compoundStatement.getId().get(0).getRoot());
+
+        return category;
+    }
+
+    private DateTimeType getDate(RCMRMT030101UK04CompoundStatement compoundStatement, ListResource parentList) {
+        if (compoundStatement != null && compoundStatement.getAvailabilityTime().getValue() != null) {
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java:127]: I think you should also check if getAvailabilityTime is not null as well

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+        }
+
+        return null;
+    }
+
+    private Meta getListMeta() {
+        return new Meta().addProfile(LIST_META_PROFILE);
+    }
+
+    private CodeableConcept getListCoding(String system, String code, String display) {
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java:151]: I'm not sure if this could be useful, but there is a `CodeableConceptUtils` class, if you'd like you can move it to the common module and use here.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java [hunk 17]
```diff
@@ -65,13 +81,21 @@ public Bundle mapToBundle(RCMRIN030000UK06Message xmlMessage) {
         var referralRequests = mapReferralRequests(ehrFolder, patient);
         addEntries(bundle, referralRequests);
 
-        var observations = mapObservations(ehrExtract, patient, List.of()); //TODO: Provide list of encounters
+        var observations = mapObservations(ehrExtract, patient, encounters);
         addEntries(bundle, observations);
 
+        // TODO: Add references to mapped resources in their appropriate lists
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java:87]: Do we have a ticket for that?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/EhrResourceExtractorUtil.java [hunk 18]
```diff
@@ -63,11 +63,11 @@ private static boolean validObservationStatement(RCMRMT030101UK04Component4 comp
         return component.getObservationStatement() != null && component.getObservationStatement().getId() == resourceId;
     }
 
-    private static boolean hasEhrComposition(RCMRMT030101UK04Component3 component) {
+    public static boolean hasEhrComposition(RCMRMT030101UK04Component3 component) {
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/EhrResourceExtractorUtil.java:66]: You can move this and `hasEhrFolder` method higher, so they're over private methods.

**Review by mrogowska — CHANGES_REQUESTED** [review]

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java (hunk 17 above)
Comment (rorysproule1) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java:87]: https://gpitbjss.atlassian.net/browse/NIAD-2051

I've added the ticket number to the comment

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+
+        var encounter = new Encounter();
+        encounter
+            .setParticipant(getEncounterParticipant(ehrComposition.getAuthor(), ehrComposition.getParticipant2()))
+            .setStatus(EncounterStatus.FINISHED)
+            .setSubject(new Reference(patient))
+            .setType(getEncounterType(ehrComposition.getCode()))
+            .setPeriod(getPeriod(ehrComposition.getEffectiveTime(), ehrComposition.getAvailabilityTime()))
+            .setIdentifier(getEncounterIdentifier(id))
+            .setMeta(getEncounterMeta())
```
Comment (boydstep) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:192]: minor - probably just `getMeta` would do (could drop "encounter" off any method name since we're in the encounter mapper class"

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 20]
```diff
@@ -0,0 +1,327 @@
...
+        } else if (CsNullFlavor.UNK.value().equals(center)) {
+            return null;
+        } else if (validValue(availabilityTimeValue)) {
+            return period.setStartElement(DateFormatUtil.parseToDateTimeType(availabilityTimeValue));
+        }
+
+        return null;
+    }
+
+    private String getTSStringValue(TS ts) {
```
Comment (boydstep) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:240]: Might want to check with Alan but you're both doing the same logic here so perhaps one of you wants to make it reusable

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 21]
```diff
@@ -0,0 +1,327 @@
...
+    private boolean validValue(String value) {
+        return value != null && !CsNullFlavor.UNK.value().equals(value);
+    }
+
+    private EncounterParticipantComponent getRecorder(RCMRMT030101UK04Author author) {
+        var recorder = new EncounterParticipantComponent();
+        var coding = new Coding();
+
+        coding
+            .setSystem(RECORDER_SYSTEM)
```
Comment (boydstep) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:273]: minor - Coding has a constructor that takes these 3 values so we could do that instead (applies below too)

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 22]
```diff
@@ -0,0 +1,327 @@
...
+        List<RCMRMT030101UK04Participant2> participant2List) {
+        List<EncounterParticipantComponent> participantList = new ArrayList<>();
+
+        if (author.getNullFlavor() == null) {
+            participantList.add(getRecorder(author));
+        }
+
+        // TODO: If author has a nullFlavor then create a recorder which references the Unknown Practitioner (NIAD-2026)
+
+        var participant2 = participant2List.stream().filter(this::isNonNullParticipant2).findFirst();
```
Comment (boydstep) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:310]: minor - an `ifPresent` could be chained here and we could do `.ifPresent(participant ->[add to whatever]);` to get rid of the separate if statement

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java (hunk 16 above)
Comment (boydstep) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java:151]: maybe a bit fussy but since Coding is an object we know/care about elsewhere a more appropriate name might be `getListCodeableConcept`

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java [hunk 23]
```diff
@@ -65,13 +81,21 @@ public Bundle mapToBundle(RCMRIN030000UK06Message xmlMessage) {
...
+        addEntries(bundle, topics);
```
Comment (boydstep) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java:88]: Is there any reason these can't be done above with the others?

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/EncounteMapperTest.java [hunk 24]
```diff
@@ -0,0 +1,334 @@
...
+    @Test
+    public void testEncountersWithMultipleCompoundStatements() {
+        var ehrExtract = unmarshallEhrExtractElement(ENCOUNTER_WITH_MULTIPLE_COMPOUND_STATEMENTS_XML);
+
+        Map<String, List<? extends DomainResource>> mappedResources = encounterMapper.mapEncounters(ehrExtract, patient);
+
+        var encounterList = mappedResources.get(ENCOUNTER_KEY);
+        assertThat(mappedResources.get(ENCOUNTER_KEY).size()).isOne();
+        assertThat(mappedResources.get(CONSULTATION_KEY).size()).isOne();
+        assertThat(mappedResources.get(TOPIC_KEY).size()).isEqualTo(2);
```
Comment (boydstep) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/EncounteMapperTest.java:107]: The style checks will want these "magic numbers" defined as constants

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/EncounteMapperTest.java (hunk 8 above)
Comment (boydstep) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/EncounteMapperTest.java:334]: We gots an extra line here bud

File: schema/src/main/java/org/hl7/v3/ObjectFactory.java [hunk 25]
```diff
@@ -3522,7 +3522,7 @@ public RCMRIN030000UK06Subject createRCMRIN030000UK06Subject() {
     }
 
     @XmlElementDecl(namespace = "urn:hl7-org:v3", name = "addr")
-    public JAXBElement<AD> createPatient(AD address) {
+    public JAXBElement<AD> createAddress(AD address) {
```
Comment (boydstep) [inline schema/src/main/java/org/hl7/v3/ObjectFactory.java:3525]: What is the reason for this change? (I mean it looks correct to be changed? But no documentation change? And we're not using it?) All the questions :D

Comment (rorysproule1) [reply schema/src/main/java/org/hl7/v3/ObjectFactory.java:3525]: Was named wrong in my last PR, it's used in testing the AddressUtil for extracting the address test xml to java object

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/EncounteMapperTest.java (hunk 24 above)
Comment (rorysproule1) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/EncounteMapperTest.java:107]: Strangely i think it still passes with 2. I've added a constant regardless.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java (hunk 23 above)
Comment (rorysproule1) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java:88]: I've put this at the bottom because we will need some functionality after mapping all the other resources to then add them to the appropriate topic and category lists in my next ticket. This might change but im not sure. For now it's the most sensible spot for it.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 26]
```diff
@@ -0,0 +1,327 @@
...
+        return ehrExtract
+            .getComponent()
+            .stream()
+            .filter(EhrResourceExtractorUtil::hasEhrFolder)
+            .map(RCMRMT030101UK04Component::getEhrFolder)
+            .map(RCMRMT030101UK04EhrFolder::getComponent)
+            .flatMap(List::stream)
+            .filter(EhrResourceExtractorUtil::hasEhrComposition)
+            .map(RCMRMT030101UK04Component3::getEhrComposition)
+            .filter(ehrComposition -> isEncounterEhrComposition(ehrComposition.getCode().getCode(), ehrComposition.getComponent()))
```
Comment (ElliotB-96) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:160]: why not just pass the ehrComposistion?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 27]
```diff
@@ -0,0 +1,327 @@
...
+    private boolean hasSuppressedContent(RCMRMT030101UK04Component4 component) {
+        return component.getEhrEmpty() != null || component.getRegistrationStatement() != null;
+    }
+
+    private List<RCMRMT030101UK04CompoundStatement> getTopicCompoundStatements(RCMRMT030101UK04EhrComposition ehrComposition) {
+        return ehrComposition
+            .getComponent()
+            .stream()
+            .map(RCMRMT030101UK04Component4::getCompoundStatement)
+            .filter(compoundStatement -> compoundStatement != null && compoundStatement.getClassCode().get(0).equals(TOPIC_CLASS_CODE))
```
Comment (ElliotB-96) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:177]: Same as above could you not create a filter method that accepts the compoundStatement

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java (hunk 16 above)
Comment (rorysproule1) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ListMapper.java:151]: As per Gosia's request, this has been moved to the CodeableConceptUtil class, so I've called it getCodeableConceptWithCoding

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java [hunk 28]
```diff
@@ -0,0 +1,327 @@
...
+    }
+
+    private List<Identifier> getEncounterIdentifier(String id) {
+        Identifier identifier = new Identifier()
+            .setSystem(IDENTIFIER_SYSTEM) // TODO: concatenate source practice org id to URL (NIAD-2021)
+            .setValue(id);
+        return List.of(identifier);
+    }
+
+    private Period getPeriod(IVLTS effectiveTime, TS availabilityTime) {
```
Comment (ElliotB-96) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:215]: Could you not pass in the ehrComposition here? it appears to be the only values you pass in

File: common/src/main/java/uk/nhs/adaptors/common/util/CodeableConceptUtils.java [hunk 29]
```diff
@@ -13,4 +14,9 @@ public static CodeableConcept createCodeableConcept(String code, String system,
             .setDisplay(display);
         return codeableConcept;
     }
+
+    public static CodeableConcept getCodeableConceptWithCoding(String system, String code, String display) {
```
Comment (mrogowska) [inline common/src/main/java/uk/nhs/adaptors/common/util/CodeableConceptUtils.java:18]: Sorry Rory, I was thinking that you could use already existing method `createCodeableConcept`. It already creates `coding`, but a bit differently by doing `getCodingFirstRep()` and then setting the code and the system. It also sets text, but you can just pass null there.

**Review by mrogowska — CHANGES_REQUESTED** [review]

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java (hunk 20 above)
Comment (rorysproule1) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:264]: Alan's has added getters comparators to this TS class, so i have used the same logic, however our logic is slightly different so there is no needs for a shared Util for this.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java (hunk 2 above)
Comment (rorysproule1) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/EncounterMapper.java:60]: As per our conversation, this isnt possible due to Consultaiton, Topic and Categories all being List Resources.

File: gp2gp-translator/src/test/resources/xml/Encounter/availability_time_encounter_period.xml [hunk 30]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bchotkowski) [inline gp2gp-translator/src/test/resources/xml/Encounter/availability_time_encounter_period.xml:1]: Just saying that for testing purposes, the <?xml> line isn't needed for Jaxb to correctly unmarshall it

**Review by bbell-kainos — APPROVED** [review]

**Review by dawidskainos — DISMISSED** [review]

**Review by bbell-kainos — DISMISSED** [review]

**Review by mrogowska — APPROVED** [review]

**Review by bbell-kainos — APPROVED** [review]
